infra/discovery/src/listener/listener.js

```diff
@@ -134,84 +136,92 @@ async function main() {
-      // then makes a purchase. If the handler would process the purchase
-      // event before the identity update, we may not be able to send the buyer
-      // a confirmation email about their offer.
-      newEvents = sortBy(newEvents, ['blockNumber', 'logIndex'], ['asc', 'asc'])
-
-      // Have the handler process each event.
-      for (const event of newEvents) {
-        // Note: we purposely do not set the exitOnError option of withRetrys to false.
-        // In case all retries fails, it indicates something is wrong at the system
-        // level and a process restart may fix it.
-        await withRetrys(async () => {
-          return handleEvent(event, context)
+      logger.debug(`Got ${newEvents.length} new events for ${contractKey}`)
+
+      if (context.config.concurrency > 1) {
+        // Concurrency greater than 1 -> Backfill mode.
+        const limit = pLimit(context.config.concurrency)
+        const promises = []
+        newEvents.forEach(newEvent => {
+          promises.push(limit(() => handleEvent(newEvent, context)))
+        })
+        await Promise.all(promises).catch(err => {
+          logger.error(`Unhandled error in event handler`)
+          logger.error(err)
         })
+      } else {
+        // Have the handler process each event.
+        for (const event of newEvents) {
+          // Note: we purposely do not set the exitOnError option of withRetrys to false.
+          // In case all retries fails, it indicates something is wrong at the system
+          // level and a process restart may fix it.
+          await withRetrys(async () => {
+            try {
+              return handleEvent(event, context)
+            } catch (err) {
+              logger.error(`Unhandled error in event handler`)
+              logger.error(err)
+            }
+          })
+        }
       }
-    }
 
-    // Record state of processing
-    logger.debug(`Updating last processed block to ${toBlock}`)
-    /**
-     * Don't assume it's the latest block known above, but use the one from
-     * event-cache
-     */
-    processedToBlock = Math.min(
-      contractsContext.marketplace.eventCache.lastQueriedBlock,
-      contractsContext.identityEvents.eventCache.lastQueriedBlock
-    )
-    await setLastBlock(context.config, processedToBlock)
-    if (context.config.enableMetrics) {
-      blockGauge.set(processedToBlock)
+      // Record state of processing
+      logger.debug(`Updating last processed block to ${toBlock}`)
+      /**
+       * Don't assume it's the latest block known above, but use the one from
+       * event-cache
+       */
+      const latestIndexedBlock =
+        contracts[contractKey].eventCache.latestIndexedBlock
+      if (processedToBlock < latestIndexedBlock) {
+        processedToBlock = latestIndexedBlock
+        await setLastBlock(context.config, processedToBlock, `${contractKey}_`)
+        if (context.config.enableMetrics) {
+          blockGauge.set(processedToBlock)
```

Comment:
Should we have a metric per contract as opposed to a single global metric. Perhaps add a TODO

Reply:
Will add TODO, since at the moment, I know nothing about prom.  Let me know if I should dig into it.